doc/src/challenges/0007-atomic-types.md

```diff
@@ -0,0 +1,126 @@
+# Challenge 7: Safety of Methods for Atomic Types & Atomic Intrinsics
+
+- **Status:** Open
+- **Tracking Issue:** [#83](https://github.com/model-checking/verify-rust-std/issues/83)
+- **Start date:** *2024-09-19*
+- **End date:** *2024-12-10*
+
+-------------------
+
+## Goal
+
+(Throughout this challenge, when we say "safe", it is identical to saying "does not exhibit undefined behavior").
+
+`core::sync::atomic` provides methods that operate on atomic types.
+For example, `atomic_store(dst: *mut T, val: T, order: Ordering)` stores `val` at the memory location pointed to by `dst` according to the specified [atomic memory ordering](https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html).
+Rust developers can use these methods to ensure that their concurrent code is thread-safe.
+
+The goal of this challenge is to verify that these methods (and the intrinsincs they invoke) are safe.
+
+### Success Criteria
+
+#### Part 1: `from_ptr` Method for Atomic Types
+
+First, verify that the unsafe `from_ptr` methods are safe.
+
+Write safety contracts for each of the `from_ptr` methods:
+
+- `AtomicBool::from_ptr`
+- `AtomicPtr::from_ptr`
+- `AtomicI8::from_ptr`
+- `AtomicU8::from_ptr`
+- `AtomicI16::from_ptr`
+- `AtomicU16::from_ptr`
+- `AtomicI32::from_ptr`
+- `AtomicU32::from_ptr`
+- `AtomicI64::from_ptr`
+- `AtomicU64::from_ptr`
+- `AtomicI128::from_ptr`
+- `AtomicU128::from_ptr`
+
+Specifically, encode the conditions about `ptr`'s alignment and validity (marked `#Safety` in the comments above the methods) as preconditions.
+Then, verify that the methods are safe for all possible values for the type that `ptr` points to, given that `ptr` satisfies those preconditions.
+
+For example, `AtomicI8::from_ptr` is defined as:
+```rust
+/// `ptr` must be [valid] ... (comment continues; omitted for brevity)
+pub const unsafe fn from_ptr<'a>(ptr: *mut i8) -> &'a AtomicI8 {
+    // SAFETY: guaranteed by the caller
+    unsafe { &*ptr.cast() }
+}
+```
+
+To verify this method, first encode the safety comments (e.g., about pointer validity) as preconditions, then verify the absence of undefined behavior for all possible `i8` values.
+
+For the `AtomicPtr` case only, we do not require that you verify safety for all possible values for the type pointed to.
+Concretely, below is the type signature for `AtomicPtr::from_ptr`:
+
+```rust
+pub const unsafe fn from_ptr<'a>(ptr: *mut *mut T) -> &'a AtomicPtr<T>
+```
+
+The type pointed to is a `*mut T`.
+Verify that for any arbitrary value for this `*mut T` (i.e., any arbitrary memory address), the method is safe.
+However, you need not verify the method for all possible instantiations of `T`.
+It suffices to verify this method for `T` of byte sizes 0, 1, 2, 4, and at least one non-power of two.
+
+#### Part 2: Unsafe Atomic Functions
```

Comment:
And/0r are you asking that that this part:

> Also write contracts enforcing that the functions are not invoked with `order`s that would cause them to panic.

Be changed so that the contracts are on the safe abstractions (the methods) instead of the unsafe methods?

Reply:
For now, I added a new section called "Safe Abstractions" that changes the contracts about panicking to be on the safe methods. LMK if I misinterpreted.

Reply:
Sorry, I'll try to be more specific next time. I was adding the comment while waiting to board my flight. ☺️

I meant that this section is named unsafe functions, but the target functions are in fact safe. Instead, this is about verifying the usage of unsafe inside the safe function. Safe functions that invoke unsafe are known as safe abstractions. Let me know if that makes sense

Reply:
Makes sense. Made some changes.